File: src/main/java/com/tesco/substitutions/commons/shutdown/ShutdownUtils.java
package com.tesco.substitutions.commons.shutdown;

import com.tesco.personalisation.commons.logging.LoggerHandler;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.eventbus.EventBus;
import io.vertx.core.json.JsonObject;
import io.vertx.rxjava.core.Vertx;
import io.vertx.rxjava.core.http.HttpServer;
import io.vertx.rxjava.ext.web.Router;
import io.vertx.rxjava.ext.web.RoutingContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import rx.Completable;
import rx.Observable;
import rx.schedulers.Schedulers;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;
import java.util.stream.Stream;

public class ShutdownUtils {
    private static final Logger LOGGER = LoggerFactory.getLogger(ShutdownUtils.class);
    private static final Integer FAILURE_THRESHOLD = 2;
    private static final Integer PERIOD_SECONDS = 2;
    private static final Integer READINESS_PROBE_DELAY;
    private static final Integer SERVER_CLOSING_DELAY;
    private static final Integer SHUTDOWN_TIMEOUT;
    private static volatile boolean isShuttingDown;
    private static volatile boolean isBeingHandled;
    private static volatile CountDownLatch gracefulDoneLatch;
    private static HttpServer httpServer;
    private static EventBus eventBus;

    public static void waitForLatch(CountDownLatch gracefulDoneLatch, Runnable completionHandler) {
        Observable.fromCallable(() -> {
            try {
                gracefulDoneLatch.await((long)SHUTDOWN_TIMEOUT, TimeUnit.MILLISECONDS);
            } catch (InterruptedException var2) {
                var2.printStackTrace();
            }

            return null;
        }).subscribeOn(Schedulers.io()).subscribe((r) -> {
            completionHandler.run();
        });
    }

    public static synchronized void handleShutdown(Vertx vertx, Runnable completionHandler) {
        isShuttingDown = true;
        if (!isBeingHandled) {
            isBeingHandled = true;
            Observable.fromCallable(() -> {
                LOGGER.info("Starting graceful shutdown");
                gracefulShutdown(vertx, gracefulDoneLatch);
                return null;
            }).subscribeOn(Schedulers.io()).subscribe();
        } else {
            LOGGER.info("Shutting down in progress");
        }

        waitForLatch(gracefulDoneLatch, completionHandler);
    }

    public static void setupReadinessLivenessProbe(Router router, String endpoint, String unhealthyResourcesMessage, Completable probeCompletable) {
        buildProbeEndpoint(endpoint, router, (rc, json) -> {
            probeCompletable.subscribe(() -> {
                json.put("health", "OK");
                rc.response().setStatusCode(200).end(json.encodePrettily());
            }, (onError) -> {
                json.put("health", "NOT OK");
                LOGGER.info(unhealthyResourcesMessage);
                rc.response().setStatusCode(500).end(json.encodePrettily());
            });
        });
    }

    public static void setupReadinessLivenessProbe(Router router, String endpoint, Handler<RoutingContext> probeHandler) {
        buildProbeEndpoint(endpoint, router, (rc, json) -> {
            probeHandler.handle(rc);
        });
    }

    private static void buildProbeEndpoint(String endpoint, Router router, BiConsumer<RoutingContext, JsonObject> handlerConsumer) {
        router.get(endpoint).handler((rc) -> {
            JsonObject json = new JsonObject();
            if (isShuttingDown) {
                LOGGER.info("Server is shutting down -- sending failure for readiness probe");
                json.put("health", "NOT OK");
                rc.response().setStatusCode(500).end(json.encodePrettily());
            } else {
                handlerConsumer.accept(rc, json);
            }

        });
    }

    private static void gracefulShutdown(Vertx vertx, CountDownLatch gracefulDoneLatch) {
        LOGGER.info("Waiting for readiness probe to fail...");
        LOGGER.info("Pending requests is: {}", LoggerHandler.getPendingRequestsCount());
        LoggerHandler.logPendingRequests();
        vertx.setTimer((long)READINESS_PROBE_DELAY, (timerId) -> {
            LOGGER.info("Stopping HTTP server and EventBus...");
            LoggerHandler.logPendingRequests();
            closeHttpServer();
            closeEventBus();
            vertx.setTimer((long)SERVER_CLOSING_DELAY, (t) -> {
                LOGGER.info("Triggering cleanup after waiting...");
                performCleanup(gracefulDoneLatch);
            });
        });
    }

    private static void waitForPendingRequests(Vertx vertx, ShutdownUtils.CommandInterface handler) {
        if (LoggerHandler.getPendingRequestsCount() != 0) {
            vertx.setPeriodic(250L, (timerId) -> {
                LOGGER.info("Checking for pending requests...");
                if (LoggerHandler.getPendingRequestsCount() == 0) {
                    LOGGER.info("No pending requests -- continue with shutdown");
                    vertx.cancelTimer(timerId);
                    handler.accept();
                }

            });
        } else {
            LOGGER.info("No pending requests -- continue with shutdown");
            handler.accept();
        }

    }

    private static void closeHttpServer() {
        if (httpServer != null) {
            httpServer.close((ar) -> {
                if (ar.succeeded()) {
                    LOGGER.info("HTTP server closed");
                } else {
                    LOGGER.error("Failed to close HTTP server", ar.cause());
                }

            });
        }

    }

    private static void closeEventBus() {
        if (eventBus != null) {
            eventBus.close((ar) -> {
                if (ar.succeeded()) {
                    LOGGER.info("Event bus closed");
                } else {
                    LOGGER.error("Failed to close event bus", ar.cause());
                }

            });
        }

    }

    private static void performCleanup(CountDownLatch countDownLatch) {
        LOGGER.info("Cleaning up resources and finalizing shutdown...");
        countDownLatch.countDown();
    }

    public void registerShutdownHook(Vertx vertx) {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            this.unDeployVerticles(vertx, new CountDownLatch(1));
        }));
    }

    private void unDeployVerticles(Vertx vertx, CountDownLatch countDownLatch) {
        Stream<String> verticleIds = vertx.deploymentIDs().stream();
        Observable.from(verticleIds.map(vertx::rxUndeploy).toArray()).doOnCompleted(() -> {
            LOGGER.info("Undeploy has been issued for all verticles");
            countDownLatch.countDown();
        }).subscribe();

        try {
            countDownLatch.await(15000L, TimeUnit.MILLISECONDS);
        } catch (InterruptedException ie) {
            ie.printStackTrace();
        }
    }

    public static void stopVerticle(Vertx vertx, Future<Void> stopFuture, JsonObject config, Runnable completionHandler) {
        if (isGracefulShutdownEnabled(config)) {
            handleShutdown(vertx, completionHandler);
        } else {
            stopFuture.complete();
        }

    }

    public static boolean isGracefulShutdownEnabled(JsonObject config) {
        return config.getBoolean("gracefulShutdownEnabled") == null ? false : config.getBoolean("gracefulShutdownEnabled");
    }

    public static void setServer(HttpServer server) {
        httpServer = server;
    }

    public static void setEventBus(EventBus coreEventBus) {
        eventBus = coreEventBus;
    }

    static {
        READINESS_PROBE_DELAY = (2 + FAILURE_THRESHOLD * PERIOD_SECONDS) * 1000;
        SERVER_CLOSING_DELAY = 5000;
        SHUTDOWN_TIMEOUT = 20000;
        isShuttingDown = false;
        isBeingHandled = false;
        gracefulDoneLatch = new CountDownLatch(1);
    }

    @FunctionalInterface
    private interface CommandInterface {
        void accept();
    }
}

File: src/main/java/com/nyala/server/common/shutdown/ShutdownUtils.kt
package com.nyala.server.common.shutdown

import com.nyala.server.common.logging.LoggerHandler
import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.eventbus.EventBus
import io.vertx.core.json.JsonObject
import io.vertx.rxjava.core.Vertx
import io.vertx.rxjava.core.http.HttpServer
import io.vertx.rxjava.ext.web.Router
import io.vertx.rxjava.ext.web.RoutingContext
import org.slf4j.LoggerFactory
import rx.Completable
import rx.Observable
import rx.schedulers.Schedulers
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.function.BiConsumer

class ShutdownUtils {

    companion object {

        private val LOGGER = LoggerFactory.getLogger(ShutdownUtils::class.java)
        private const val FAILURE_THRESHOLD = 2
        private const val PERIOD_SECONDS = 2
        private var READINESS_PROBE_DELAY: Long? = null
        private var SERVER_CLOSING_DELAY: Long? = null
        private var SHUTDOWN_TIMEOUT: Long? = null

        @Volatile
        private var isShuttingDown = false

        @Volatile
        private var isBeingHandled = false

        @Volatile
        private var gracefulDoneLatch: CountDownLatch? = null
        private var httpServer: HttpServer? = null
        private var eventBus: EventBus? = null

        private fun waitForLatch(gracefulDoneLatch: CountDownLatch?, completionHandler: Runnable) {
            Observable.fromCallable {
                try {
                    gracefulDoneLatch!!.await(SHUTDOWN_TIMEOUT as Long, TimeUnit.MILLISECONDS)
                } catch (exception: InterruptedException) {
                    exception.printStackTrace()
                }
                null
            }.subscribeOn(Schedulers.io()).subscribe { completionHandler.run() }
        }

        @Synchronized
        fun handleShutdown(vertx: Vertx, completionHandler: Runnable) {
            isShuttingDown = true
            if (!isBeingHandled) {
                isBeingHandled = true
                Observable.fromCallable {
                    LOGGER.info("Starting graceful shutdown")
                    gracefulShutdown(vertx, gracefulDoneLatch)
                    null
                }.subscribeOn(Schedulers.io()).subscribe()
            } else {
                LOGGER.info("Shutting down in progress")
            }
            waitForLatch(gracefulDoneLatch, completionHandler)
        }

        fun setupReadinessLivenessProbe(router: Router, endpoint: String, unhealthyResourcesMessage: String?, probeCompletable: Completable) {
            buildProbeEndpoint(endpoint, router)
            { rc: RoutingContext, json: JsonObject ->
                probeCompletable.subscribe({
                    json.put("health", "OK")
                    rc.response().setStatusCode(200).end(json.encodePrettily())
                }) {
                    json.put("health", "NOT OK")
                    LOGGER.info(unhealthyResourcesMessage)
                    rc.response().setStatusCode(500).end(json.encodePrettily())
                }
            }
        }

        fun setupReadinessLivenessProbe(router: Router, endpoint: String, probeHandler: (RoutingContext) -> Unit) {
            buildProbeEndpoint(endpoint, router) {
                rc: RoutingContext, _: JsonObject ->
                probeHandler(rc)
            }
        }

        private fun buildProbeEndpoint(endpoint: String, router: Router, handlerConsumer: (RoutingContext, JsonObject) -> Unit) {
            router[endpoint].handler { rc: RoutingContext ->
                val json = JsonObject()
                if (isShuttingDown) {
                    LOGGER.info("Server is shutting down -- sending failure for readiness probe")
                    json.put("health", "NOT OK")
                    rc.response().setStatusCode(500).end(json.encodePrettily())
                } else {
                    handlerConsumer(rc, json)
                }
            }
        }

        private fun gracefulShutdown(vertx: Vertx, gracefulDoneLatch: CountDownLatch?) {
            LOGGER.info("Waiting for readiness probe to fail...")
            LOGGER.info("Pending requests is: {}", LoggerHandler.getPendingRequestsCount())
            LoggerHandler.logPendingRequests()
            vertx.setTimer(READINESS_PROBE_DELAY as Long) {
                LOGGER.info("Stopping HTTP server and EventBus...")
                LoggerHandler.logPendingRequests()
                closeHttpServer()
                closeEventBus()
                vertx.setTimer(SERVER_CLOSING_DELAY as Long) {
                    LOGGER.info("Triggering cleanup after waiting...")
                    performCleanup(gracefulDoneLatch)
                }
            }
        }

        private fun closeHttpServer() {
            if (httpServer != null) {
                httpServer!!.close { ar: AsyncResult<Void?> ->
                    if (ar.succeeded()) {
                        LOGGER.info("HTTP server closed")
                    } else {
                        LOGGER.error("Failed to close HTTP server", ar.cause())
                    }
                }
            }
        }

        private fun closeEventBus() {
            if (eventBus != null) {
                eventBus!!.close { ar: AsyncResult<Void?> ->
                    if (ar.succeeded()) {
                        LOGGER.info("Event bus closed")
                    } else {
                        LOGGER.error("Failed to close event bus", ar.cause())
                    }
                }
            }
        }

        private fun performCleanup(countDownLatch: CountDownLatch?) {
            LOGGER.info("Cleaning up resources and finalizing shutdown...")
            countDownLatch!!.countDown()
        }

        @JvmStatic
        fun stopVerticle(vertx: Vertx, stopFuture: Future<Void?>, config: JsonObject, completionHandler: Runnable) {
            if (isGracefulShutdownEnabled(config)) {
                handleShutdown(vertx, completionHandler)
            } else {
                stopFuture.complete()
            }
        }

        private fun isGracefulShutdownEnabled(config: JsonObject): Boolean {
            return if (config.getBoolean("gracefulShutdownEnabled") == null) false else config.getBoolean("gracefulShutdownEnabled")
        }

        fun setServer(server: HttpServer?) {
            httpServer = server
        }

        @JvmStatic
        fun setEventBus(coreEventBus: EventBus?) {
            eventBus = coreEventBus
        }

        init {
            READINESS_PROBE_DELAY = (2 + FAILURE_THRESHOLD * PERIOD_SECONDS) * 1000L
            SERVER_CLOSING_DELAY = 5000
            SHUTDOWN_TIMEOUT = 20000
            isShuttingDown = false
            isBeingHandled = false
            gracefulDoneLatch = CountDownLatch(1)
        }
    }

    fun registerShutdownHook(vertx: Vertx) {
        Runtime.getRuntime().addShutdownHook(Thread(Runnable { unDeployVerticles(vertx, CountDownLatch(1)) }))
    }

    private fun unDeployVerticles(vertx: Vertx, countDownLatch: CountDownLatch) {
        val verticleIds = vertx.deploymentIDs().stream()
        Observable.from(verticleIds.map { deploymentID: String? -> vertx.rxUndeploy(deploymentID) }.toArray()).doOnCompleted {
            LOGGER.info("Undeploy has been issued for all verticles")
            countDownLatch.countDown()
        }.subscribe()
        try {
            countDownLatch.await(15000L, TimeUnit.MILLISECONDS)
        } catch (ie: InterruptedException) {
            ie.printStackTrace()
        }
    }
}